include/zephyr/arch/riscv/atomic.h

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) 2024 NextSilicon
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_ARCH_RISCV_ATOMIC_H_
+#define ZEPHYR_INCLUDE_ARCH_RISCV_ATOMIC_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* The standard RISC-V atomic-instruction extension, "A", specifies
+ * the number of instructions that atomically read-modify-write memory,
+ * which RISC-V harts should support in order to synchronise harts
+ * running in the same memory space. This is the subset of RISC-V
+ * atomic-instructions not present in atomic_builtin.h file.
+ */
+
+#ifdef CONFIG_64BIT
+static ALWAYS_INLINE atomic_val_t atomic_swap(const atomic_t *target, atomic_val_t newval)
+{
+	atomic_val_t ret;
+
+	__asm__ volatile("amoswap.d.aq  %0, %1, %2"
+			 : "=r"(ret)
+			 : "r"(newval), "A"(*target)
+			 : "memory");
+
+	return ret;
+}
+
+static ALWAYS_INLINE atomic_val_t atomic_lrsc_cas(atomic_t *target, atomic_val_t oldval,
+					     atomic_val_t newval)
+{
+	atomic_val_t ret;
+
+	__asm__ volatile("lr.d.aq %0, %3\n"
+			"bne %0, %1, 1f\n"
+			"sc.d.aqrl %0, %2, %3\n"
+			"1:\n"
+			: "=&r"(ret)
+			: "r"(oldval), "r"(newval), "A"(*target)
+			: "memory");
+
+	return ret == 0;
+}
+
+static ALWAYS_INLINE atomic_val_t atomic_max(atomic_t *target, atomic_val_t value)
+{
+	atomic_val_t ret;
+
+	__asm__ volatile("amomax.d.aq  %0, %1, %2"
+			 : "=r"(ret)
+			 : "r"(value), "A"(*target)
+			 : "memory");
+
+	return ret;
+}
+
+static ALWAYS_INLINE atomic_val_t atomic_min(atomic_t *target, atomic_val_t value)
+{
+	atomic_val_t ret;
+
+	__asm__ volatile("amomin.d.aq  %0, %1, %2"
+			 : "=r"(ret)
+			 : "r"(value), "A"(*target)
+			 : "memory");
+
+	return ret;
+}
+
+#else /* CONFIG_64BIT */
+
+static ALWAYS_INLINE atomic_val_t atomic_swap(const atomic_t *target, atomic_val_t newval)
+{
+	atomic_val_t ret;
+
+	__asm__ volatile("amoswap.w.aq  %0, %1, %2"
+			 : "=r"(ret)
+			 : "r"(newval), "A"(*target)
+			 : "memory");
+
+	return ret;
+}
+
+static ALWAYS_INLINE atomic_val_t atomic_lrsc_cas(atomic_t *target, atomic_val_t oldval,
+					     atomic_val_t newval)
```

Comment:
What's the advantage of this vs. the compiler emitted one, which AFAICT looks the same.  An atomic_cas() in the current tree emits this for qemu_riscv32, for example, which looks broadly identical:

```
        return __atomic_compare_exchange_n(target, &old_value, new_value,
80000760:       07b00693                li      a3,123
80000764:       01878793                addi    a5,a5,24
80000768:       0f50000f                fence   iorw,ow
8000076c:       1407a52f                lr.w.aq a0,(a5)
80000770:       00e51563                bne     a0,a4,8000077a <main+0x42>
80000774:       1cd7a62f                sc.w.aq a2,a3,(a5)
80000778:       fa75                    bnez    a2,8000076c <main+0x34>
```

Reply:
Right, removed it.